src/mr/default_memory_resource.cpp

```diff
@@ -14,23 +14,46 @@
  * limitations under the License.
  */
 #include <rmm/mr/cuda_memory_resource.hpp>
+#include <rmm/mr/cnmem_memory_resource.hpp>
+#include <rmm/mr/managed_memory_resource.hpp>
 #include <rmm/mr/default_memory_resource.hpp>
 #include <rmm/mr/device_memory_resource.hpp>
 
 #include <atomic>
 
 namespace rmm {
 namespace mr {
-namespace {
-// The initial, default memory resource is a `cuda_memory_resource`
+
+namespace detail{
 device_memory_resource* initial_resource() {
+  return cuda_resource();
+}
+
+device_memory_resource* cuda_resource() {
   static cuda_memory_resource resource{};
   return &resource;
 }
 
+device_memory_resource* pool_resource(std::size_t pool_size){
+   static cnmem_memory_resource mr{pool_size};
+   return &mr;
+}
+
+
+device_memory_resource* managed_resource(){
+   static managed_memory_resource mr{};
+   return &mr;
+}
+} //namespace detail
+namespace {
+
+
+
+
+
 // Use an atomic to guarantee thread safety
 std::atomic<device_memory_resource*>& get_default() {
```

Comment:
Don't you need thread safe ways to get the pool_resource, managed_resource, etc?

Reply:
Singletons in C++11 and beyond are always threadsafe (only one thread is guaranteed to construct the static object). And since they never change (i.e., there is no way to modify what is returned by `pool_resource()`), there's no need to provide any additional thread safety mechanism.

Reply:
OK, that wasn't my question. My question was if we have this way to get the default MR, don't we need equivalents for the pool_resource and managed_resource MR accessors?

I guess what you are saying is this function doesn't need to return an atomic.

Reply:
The only reason `get_default` returns an `atomic` is because there is a corresponding `set_default` function. If two threads called `set_default` and `get_default` concurrently, we need to protect against race conditions via an `atomic`.

In contrast, with `pool_resource()`, there's no chance for race conditions, so no need for an `atomic`.

Reply:
Ah, I see.